:sparkles: feat: 매거진 수정안 링크 전송 기능 구현 및 매거진 미리보기 토큰 관리 검증 기능 추가

jiyoon0000 opened the pull request:
-관리자가 로컬 크리에이터에게 매거진 수정본 확인 요청 메일을 전송하는 기능 구현
-revisionCount에 따라 메일 제목/내용 분기 처리
-ErrorCode Enum에 피드백 전용 예외코드 추가
  -FEEDBACK_NOT_FOUND (피드백 없음)
  -FEEDBACK_ALREADY_REFLECTED (이미 반영된 피드백)
-피드백이 있을 시에만 수정할 수 있는 로직으로 변경
-추후 오타나 관리자의 실수로 인한 수정 시 revisionCount와 피드백 상관없이 수정할 수 있는 API 추가 예정
-초안/수정안 발송 시 기존 토큰 무효화 후 재발급
-미리보기 접근 시 validatePreviewToken() 통해 만료·본인 검증

<img width="881" height="602" alt="스크린샷 2025-09-12 오전 1 42 42" src="https://github.com/user-attachments/assets/a88f0a93-4afe-4329-af03-95049a18d84b" />


<img width="881" height="602" alt="스크린샷 2025-09-12 오전 12 14 28" src="https://github.com/user-attachments/assets/303ee41e-43c6-425e-aec6-fce3466f2cc0" />

기쁨의 눈물......😭🫢🥹

jiyoon0000 committed:
:sparkles: feat : 매거진 도메인 엔티티 설계, 관리자 임시등록 API, 카테고리 자동 등록(CommandLineRunner) 구현
-Magazine 엔티티/DTO/enum/Repository 설계 및 생성
-RegionCategory, AptitudeCategory 엔티티/레포지토리 추가 및 초기 데이터 자동 등록(CommandLineRunner)
-엔티티 생성자 추가, 중복 체크 로직 적용
-개발 환경(local)에서만 자동 등록되도록 profile 분기 처리
-Swagger 문서화 및 테스트용 http 파일 작성

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
:sparkles: feat : 매거진 파일 업로드 구현
-매거진 썸네일/상세이미지 파일 다중 업로드
-썸네일(THUMBNAIL) 업로드 시 1개만 허용하는 예외처리 추가
-S3Uploader 연동 및 파일/파일매핑 엔티티 저장 로직 구현
-파일 목적별 업로드 분기 처리(THUMBNAIL/DETAIL_IMAGE)
-임시 업로드(referenceId=0L) 구조 적용, 실제 매거진 등록 시(사용자에게 노출) 매핑 예정

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
:sparkles: feat : 관리자 임시등록(DRAFT) 매거진 목록 조회 기능 구현
-magazineStatus=DRAFT, deletedStatus=DISPLAYED 조건의 1차등록 매거진 목록 조회
-제목, 카테고리명, 썸네일, 등록일 등 주요 정보 리스트 반환
-현재는 List 반환, 추후 페이지네이션/정렬 확장 예정
-Swagger 문서화 및 테스트용 http 파일 작성

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
:building_construction: chore : SwaggerConfig에 팀 이메일 첨부

jiyoon0000 committed:
:sparkles: feat : 임시등록(DRAFT) 매거진 상세조회 및 파일 자동매핑 기능 추가
-관리자 전용 1차 임시 매거진 상세조회 구현
(제목, 본문, 썸네일, 지역/적성 카테고리, 상세 이미지 리스트, 등록자, 생성일 반환)
-매거진 등록 시 본문 내 이미지(src) S3 URL 파싱 후,해당 파일(file_mapping)의 referenceId를 매거진ID로 자동 업데이트
-임시등록 단계에서 파일과 매거진의 관계 정확히 매핑되도록 개선
-불필요/미연결 이미지 파일 참조 이슈 방지 및 데이터 정합성 강화
-DRAFT 매거진이 여러 개일 때도 각각 매핑되도록 처리
-Swagger 문서화 및 테스트용 http 파일 작성
-프론트 구현 시 파일 매핑 테스트 한번 더 필요

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
:wrench: add : 관리자 임시등록 매거진 목록 조회에 페이지네이션 적용
-기존 List<Magazine> 반환 로직을 Page<Magazine>으로 변경
-Pageable을 활용한 페이징 처리 도입

jiyoon0000 committed:
:sparkles: feat : 매거진 초안 확인 링크 이메일 전송 기능 구현
-매거진 생성 시 연결된 로컬 크리에이터에게 초안 확인 링크 전송
-매거진 생성 시 로컬 크리에이터 아이디 추가하도록 수정
-이메일 본문에 미리보기 링크 포함
-MagazinePreviewToken 엔티티로 토큰 생성 및 유효성 검증
-초안 이메일은 1회만 전송되도록 제어

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

# Conflicts:
#	src/main/java/com/salayo/locallifebackend/domain/magazine/service/MagazineService.java

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
:wrench: add : 매거진 초안 안내 메일 내용 수정 (피드백 버튼 안내 및 수정 횟수 명시)

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
:sparkles: feat : 관리자 매거진 수정 기능 구현
-MagazineRevision 엔티티 기반으로 수정 이력 저장
-수정 요청 3회 제한 검증 로직 추가

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
:sparkles: feat : 매거진 수정 이력 관리 기능 추가
-관리자가 매거진 본문을 수정할 수 있도록 구현
-수정 시 MagazineRevision에 히스토리 저장

jiyoon0000 committed:
Merge branch 'develop' into feature/magazine

jiyoon0000 committed:
:sparkles: feat : 매거진 수정안 링크 전송 기능 구현
-관리자가 로컬 크리에이터에게 매거진 수정본 확인 요청 메일을 전송하는 기능 구현
-revisionCount에 따라 메일 제목/내용 분기 처리

jiyoon0000 committed:
:recycle: refactor : 매거진 피드백/수정 관련 에러코드 분리 및 추가
-ErrorCode Enum에 피드백 전용 예외코드 추가
  -FEEDBACK_NOT_FOUND (피드백 없음)
  -FEEDBACK_ALREADY_REFLECTED (이미 반영된 피드백)
-피드백이 있을 시에만 수정할 수 있는 로직으로 변경
-추후 오타나 관리자의 실수로 인한 수정 시 revisionCount와 피드백 상관없이 수정할 수 있는 API 추가 예정

jiyoon0000 committed:
:sparkles: feat: 매거진 미리보기 토큰 관리 및 검증 기능 추가
-초안/수정안 발송 시 기존 토큰 무효화 후 재발급
-미리보기 접근 시 validatePreviewToken() 통해 만료·본인 검증

jiyoon0000 committed:
:recycle: refactor : 매거진 수정, 생성 로직 보완
-수정안 전송 시 revisionCount가 1 이상일 경우에만 동작하도록 검증 추가
-매거진 생성 시 승인된 로컬 크리에이터(APPROVED)만 등록 가능하도록 검증 추가

sooyeoneo commented on src/main/java/com/salayo/locallifebackend/domain/magazine/service/MagazineService.java:
```diff
@@ -178,14 +191,21 @@ public void sendPreviewLink(Long magazineId) {
+        if (lastFeedback.isReflected()) {
+            throw new CustomException(ErrorCode.FEEDBACK_ALREADY_REFLECTED);
+        }
+
```
선택사항:

현재 MagazineService에서 피드백 반영 상태 변경 `lastFeedback.markAsReflected()`와 매거진 내용 업데이트 magazine.updateContent() 로직이 분리되어 있는 것으로 보여집니다.

이 두 로직은 '매거진 수정' 이라는 하나의 비즈니스 행위이므로, 
Magazine 엔티티에 revisionCount 필드를 추가한 후,

`applyFeedback(lastFeedback, newContent, revisionRepository)`와 같은 메서드를 추가하여 
도메인 로직의 응집도를 높이는 방법도 있을 것 같습니다. 🤓

`MagazineService`  서비스
```
        magazine.applyFeedback(lastFeedback, magazineUpdateRequestDto.getContent(), magazineRevisionRepository);
```

`Magazine` 엔티티 
```
public void applyFeedback(MagazineFeedback feedback, String newContent, MagazineRevisionRepository revisionRepository) {
        if (feedback == null) {
            throw new CustomException(ErrorCode.FEEDBACK_NOT_FOUND);
        }
        if (feedback.isReflected()) {
            throw new CustomException(ErrorCode.FEEDBACK_ALREADY_REFLECTED);
        }

        this.content = newContent;

        feedback.markAsReflected();

        this.revisionCount += 1; // Magazine 엔티티에 revisionCount 필드가 있다면 직접 업데이트 가능
        MagazineRevision revision = new MagazineRevision(this, newContent, this.revisionCount);
        revisionRepository.save(revision);
    }        
```

jiyoon0000 replied:
수연님 코멘트 감사합니다! ☺️

매거진 내용 수정과 관련해서 `updateContent` 메서드를 둔 이유는 매거진 수정 시 

1. 로컬 크리에이터의 피드백을 반영해서 수정
2. 관리자의 실수(오타나 문단변경 등)로 인한 수정 

두가지 케이스로 나눠질 수 있다고 생각하고 짠 로직인데요, 
메서드를 분리하게 된다면 같은 기능을 하는 메서드가 두개가 생기는 것이라 따로 분리하지 않고 공통으로 가져가는 것이 좋을 것 같다고 생각했습니다.

또한, 제안해주신 엔티티에 메서드를 두는 방식은 좋은 것 같으나 Entity 단에서 Repository에 직접 접근하거나 예외를 던지는 것은 DDD 원칙에 위배된다고 보았습니다. 그래서 엔티티는 상태 변경만 담당하고 예외처리에 대한 로직은 서비스단에서 하는 방향으로 유지하는 게 좋을 것 같아요🥲

sooyeoneo commented on src/main/java/com/salayo/locallifebackend/domain/magazine/service/MagazineService.java:
```diff
@@ -163,6 +173,9 @@ public void sendPreviewLink(Long magazineId) {
         magazine.updateStatus(MagazineStatus.PENDING_CONFIRMATION);
 
+        magazinePreviewTokenRepository.deleteByMagazineId(magazineId);
+        magazinePreviewTokenRepository.flush();
```
선택 사항: 

현재 `sendPreviewLink()`와 `sendRevisionLink()` 메서드에서 
'기존 코드 삭제 → 새 토큰 생성 → 새 토큰 저장' 로직이 반복되는 것 같아 보입니다. 
이 로직을 별도의 private 메서드로 분리하면 코드 중복을 제거하고 가독성이 높아질 수 있을 것 같습니다.🥾

**private 메서드로 분리한 예시**
```
private MagazinePreviewToken reissueAndSaveToken(Magazine magazine) {
     magazinePreviewTokenRepository.deleteByMagazineId(magazine.getId());
     magazinePreviewTokenRepository.flush();

     String token = UUID.randomUUID().toString();
     LocalDateTime expiresAt = LocalDateTime.now().plusDays(3);
     String email = magazine.getLocalCreator().getMember().getEmail();

     MagazinePreviewToken newToken = new MagazinePreviewToken(magazine, email, token, expiresAt);
     return magazinePreviewTokenRepository.save(newToken);
 }
 
 ```
`sendPreviewLink()`와 `sendRevisionLink()`에서 `reissueAndSaveToken()` 메서드를 호출하여 
간결하게 사용이 가능합니다. `reissueAndSaveToken(magazine)`

jiyoon0000 replied:
초안/수정안/최종안의 링크 만료일이 달라 공통 메서드를 만드려다 말았는데, 파라미터로 만료일을 받으면 될 것 같네용!!

공통 메서드를 만들어서 사용하는 방식으로 수정해볼게요 감사합니다! 😎

sooyeoneo commented on src/main/java/com/salayo/locallifebackend/domain/magazine/service/MagazineService.java:
```diff
@@ -196,5 +216,69 @@ public void updateMagazine(Long magazineId, MagazineUpdateRequestDto magazineUpd
+        int revisionCount = magazineRevisionRepository.countByMagazineId(magazineId);
+        if (revisionCount == 0) {
+            throw new CustomException(ErrorCode.REVISION_NOT_AVAILABLE);
+        }
```
선택사항:

`sendPreviewLink()` 메서드와 마찬가지로, 이 메서드에 있는 '기존 코드 삭제 → 새 토큰 생성 → 새 토큰 저장' 로직도
별도의 private 메서드 `reissueAndSaveToken()`을 사용하면 중복 제거가 가능합니다.

jiyoon0000 replied:
네엥 반영할게욥!!

sooyeoneo reviewed (approved):
PR 코드만 보고 생각한 것이라, 
로직을 자세히는 알지 못하여 다른 의도가 있으시다면 그냥 넘어가시면 될 듯 합니다!

jiyoon0000 committed:
:recycle: refactor: 매거진 미리보기 토큰 발급 로직 공통 메서드화
-sendPreviewLink / sendRevisionLink 중복 제거
-reissueAndSaveToken(magazine, daysToExpire) 메서드 추가하여 초안(14일), 수정안(3일) 등 만료일을 파라미터로 전달

j-hann reviewed (approved):
굿굿👍
